### DISCUSSION TITLE
Terrible DX while writing seed.ts, is it possible to generate a seed file for 'complex' prisma schema's?

### DISCUSSION THREAD

**Discussion by Jarrodsz** [prisma/prisma#16854] [Q&A]
New to prisma and typescript I created my schema.prisma fairly soon, not that hard.
BUT 
Creating a seed file to get dummy data *into* the database is a absolute terrible experience!

Is there no such thing as generate a seed file from a prisma schema? Since prisma knows how the relation of the database is, it wouden't be that hard to generate sample data based on it?

For example consider these models:
```

model Organisation {
  id   String @unique @default(cuid())
  name String

  // Association to the User model,
  // indicating that each organisation can have one or more users
  users User[]

  // Association to the OrganisationProfile model,
  // indicating that an organisation can have one or more teams
  teams OrganisationTeam[]

  OrganisationProfile    OrganisationProfile[]
  OrganisationTeamInvite OrganisationTeamInvite[]
  organisationId         String

  createdAt DateTime? @default(now())
  updatedAt DateTime? @updatedAt
}

model UserSubscription {
  id String @id @default(cuid())

  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  userId String @unique

  customerId     String? @unique
  subscriptionId String?
  status         String?

  currentPeriodStart Int?
  currentPeriodEnd   Int?
  cancelAtPeriodEnd  Boolean? @default(false)

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

model User {
  id     String @id @default(cuid())
  userId String @unique

  email String @unique

  password UserPassword?

  subscription UserSubscription?

  organisation Organisation? @relation(fields: [organisationId], references: [id])

  //organisationTeamInvites  OrganisationTeamInvite[]

  // Association to the Answer model
  //answers Answer[] // MERGE IN LATER

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  OrganisationTeam       OrganisationTeam?        @relation(fields: [organisationTeamId], references: [id])
  organisationTeamId     String?
  OrganisationTeamInvite OrganisationTeamInvite[]
  organisationId         String?
  UserProfile            UserProfile?
}

model UserPassword {
  hash String

  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  userId String @unique

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

model UserProfile {
  id String @unique

  firstName String
  lastName  String

  user User @relation(fields: [id], references: [id])

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

I have cheated a bit and used ChatGP to write some code for the seeding. Schema is working fully and the prisma is generated. 

But im really not happy with having to write this kind of code:

```async function createUser(
  userId: string,
  email: string,
  password: string
  // subscription?: UserSubscription,
  // organisation?: { id: string },
  // organisationTeam?: { id: string },
  // organisationTeamInvites?: { id: string }[],
  // userProfile?: { id: string }
): Promise<User> {
  const hashedPassword = await bcrypt.hash(password, 10);
  const user: Promise<User> = await prisma.user.create({
    data: {
      userId,
      email,
      password: {
        create: {
          hash: hashedPassword
        }
      }
      // subscription: subscription ? { connect: subscription } : undefined,
      // organisation: organisation ? { connect: { id: organisation.id } } : undefined,
      // organisationTeam: organisationTeam ? { connect: { id: organisationTeam.id } } : undefined,
      // organisationTeamInvites: organisationTeamInvites
      //   ? { connect: organisationTeamInvites.map(invite => ({ id: invite.id })) }
      //   : undefined,
      //userProfile: userProfile ? { connect: { id: userProfile.id } } : undefined
    }
  });

  return user;
}
```

This above is not even working, I have some issues with ```TSError: ⨯ Unable to compile TypeScript:
prisma/seed.ts:26:9 - error TS2739: Type 'User' is missing the following properties from type 'Promise<User>': then, catch, finally, [Symbol.toStringTag]

26   const user: Promise<User> = await prisma.user.create({``` 

This Discussion is more a way to ask the community if I cannot autogenerate the seed file somehow, or else use a different more easy way to seed the database?

Since What im only want is some example data user, with user profile, subscription etc.
It shouden't be that of a dreadful experience coding for days to get some simple models in the database filled?

**Comment by janpio** [comment]
Interesting idea.

Can you describe in natural language what your seed script should do in the end?

We did have some tool to generate test data in the past, but that was really hard to maintain and not popular - so we did not keep maintaining it. Internally we also have some other tools that help with seeding, but those have not been asked for so we never "productized" them. Maybe your ask could be the push to look at this.

**Reply by Jarrodsz** [reply]
How about a prisma method in typescript implementing a method that generated::

1. createModel(arguments)
2. Write an example on how to call this method.

This way you can make the code consistant across all create methods.

Im not sure how this would work with associations though since ut would need to take into consideration all existing models(??)

**Reply by janpio** [reply]
So a function to output code that you can copy/paste into a file? 

I think that would be very much possible via a custom generator. https://www.npmjs.com/package/prisma-docs-generator comes to mind, which already generates _some_ example code but is missing the fields there that I think you would want.

**Reply by Jarrodsz** [reply]
I will inject the methods and code in the appropriate places into my seed.ts using a template to generate it. 
This is asesome didn't knew it exits. I can definitely copy some code from this thank you.

I would suggest a 'generate seed file' method for prisma-client. 
Since that probably will take a long time to be considered/included I will build a working solution for me now and stick with that or migrate if prisma could include this in future.

1.  load schema.prisma
2. generate a seed file with example methods to populate each model
3. use a template ( I called mine _seed.ts) and populate the imports from prismaclient, models, insert the create methods on the right place, insert the calls to these methods in the right place

**Reply by janpio** [reply]
Then I think a generator that you add to your Prisma schema file, and which recreated and updates this stub template with each `prisma generate` sounds like a good idea. It can just write any files into a specific folder, and then you can either extend those or copy/paste them into a new file and adapt them. If the generated file changes, you will know that you need to update your adaptation of it.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
